### ISSUE TITLE
Package Monero for Debian so it can eventually also be added to projects like Tails

### ISSUE THREAD

**Issue by scottAnselmo** [monero-project/monero#2395]
Currently Monero isn't packaged for Debian which is a [roadblock to Monero being adopted by Tails.](https://labs.riseup.net/code/issues/14390).

At time of writing [it looks like there are no packages being tested or otherwise being worked on](https://packages.debian.org/search?keywords=monero&searchon=names&suite=stable&section=all) for Debian.

**Comment by dEBRUYNE-1** [comment]
+enhancement

**Comment by privacycat** [comment]
+enhancement
Agree. Monero is much more private and anonymous than Bitcoin. It's a no-brainer to package it for Debial so that ic could be adopted by Tails.

**Comment by anonimal** [comment]
A debian package would be nice, we do have [snap](https://snapcraft.io/) ready to go though.

Note: re: tails, I don't see how running monero nodes *at large* over Tor would be a good thing for the monero network. This is why we have kovri. I'm assuming the demand is more for monero wallet + a remote node onion service?

**Comment by favioflamingo** [comment]
If a debian package is created, is it possible to make sure that there is a monero-dev and libmonero package created as well?  This way, other people can use the c++ funtions in external programs.  I have just tried to use Monero as a library, but the examples only show how to work with static libs compiled with cmake.

**Comment by hyc** [comment]
What exactly does "packaging for Debian" (or any other distro) actually mean? Do we have to actually provide scripts for their build system to build Monero binaries from source? Or are we simply providing the official getmonero.org binaries in a dpkg archive as opposed to a tar archive?

I assume the latter, because having large numbers of non-official binaries out there sounds like a security risk.

**Comment by favioflamingo** [comment]
Speaking humbly, for debian packaging, you start with a source package.  Given a source package, someone would be able to type `apt-get source -b monero` and get several debian packages with compiled binaries as a result.  The /bin scripts would get put into a monero-cli kind of package, the libraries (-fPIC flag?) would get put into a libxmr0 package, and etcetera.  Having dynamically linked libraries would be nice to have available in a package.

**Comment by diederikdehaas** [comment]
> What exactly does "packaging for Debian" (or any other distro) actually mean? Do we have to actually provide scripts for their build system to build Monero binaries from source? Or are we simply providing the official getmonero.org binaries in a dpkg archive as opposed to a tar archive?
> I assume the latter, because having large numbers of non-official binaries out there sounds like a security risk.

It is certainly **not** the latter. Debian will **never** distribute a binary that isn't build from source on their infrastructure.
It comes down to populating a `debian` directory which instructs the build system how to produce the package (so in a way the former). Then you'd have to file an `ITP` (Intention to Package) bug and find a sponsor (someone who's already a Debian Developer) to sponsor your package and when ready (s)he will upload it to the NEW queue after which an FTP master reviews the package (again) and if it meets the requirements accepts it into the archive.

Further reading:
- https://www.debian.org/doc/manuals/maint-guide/index.en.html
- https://wiki.debian.org/PackagingWithGit
- https://honk.sigxcpu.org/piki/projects/git-buildpackage/

But it's probably best to (first) contact [Debian Bitcoin Packaging Team](https://lists.alioth.debian.org/cgi-bin/mailman/listinfo/pkg-bitcoin-devel) if someone intends to do that.
(Yes, I know monero != bitcoin, but wrt packaging it may be useful to put it under 1 umbrella team)

**Comment by favioflamingo** [comment]
If someone can add a configuration in Cmake to make dynamically linked libs (soname=x.y.z, etc), then I could set up the git-buildpackage items to make it easy to compile debian packages in one command (pbuilder).  You can see my work on [the picocoin debian package](https://github.com/favioflamingo/picocoin) as an example, though picocoin works with autotools as opposed to Cmake.  I am trying to get picocoin into Debian repos at the moment as well.

**Comment by dma** [comment]
Hey all,

I've made a stab at packaging monerod and monero GUI wallet for @Subgraph OS. By default, all traffic exiting from Subgraph OS is over Tor, in case you weren't aware -- unless the user is using the sandboxed clearnet browser. So Monero users on SGOS would do so over Tor.

I just put downloadable packages on our website for testing, if all goes well, we'll drop them into our repo so users can apt-get install monero & they're good to go.  

Download links / instructions for testing before we drop them into our repo:

https://subgraph.com/sgos/documentation/monero/

The commits of the packaging metadata:

Monero:

https://github.com/subgraph/monero/commit/8ef7d55475780e6b844a5d2376bf6809a4de986e

Monero Core:

https://github.com/subgraph/monero-core/commit/b63ac3aad2ad06b43b56945dd0a3114a76904262

**Comment by dma** [comment]
As the above was really a first try mostly to re-implement "build.sh" as debian build rules, I've removed the old repos, and we're now re-doing the packages to build with pbuilder (as we do our other packages). On-going work is here in the "debian" branch of the below forked repos:

https://github.com/subgraph/monero
https://github.com/subgraph/monero-gui

Monero packages have been added to our master packages repo, but it's not building in travis-ci yet:

https://github.com/subgraph/subgraph-debian-packages

I'll update this thread when we have a final packaging we like. Almost there..

**Comment by CameronRuggles** [comment]
How is this progressing? Getting it added to TAILS is an exceedingly important project.
What currently needs to be done for Monero GUI to be added to the Debian packages?

**Comment by dma** [comment]
Oh, I packaged it in a nice way several weeks ago, but didn't want to release it or announce it until had binary reproducibility (who would want to use a wallet packaged by a third party that isn't?). 

I'd say I got almost there, but ran into some linker weirdness and then had to stop to work on something else. Also, not many people seemed to care and I had other priorities. 

I'd recommend Subgraph over Tails for storing and using XMR, but that's just me :)

I'll be back on it soon and return to this thread when I have achieved byte-for-byte matches on the final .debs.

If you want to try our packages in Tails, you're welcome to:

https://subgraph.com/sgos/documentation/monero/

Those date from when monero-gui was still monero-core, I'll build some new ones and update the page sometime in the next few days.

**Comment by sanecito** [comment]
Cool, keep us updated! I'll test it out on Tails and Subgraph when you put out the new builds and send some pizza/beer money your way from Tails or Subgraph. Will probably advertise it on the various Monero forums as well so you can get more people to test Monero (and Subgraph by proxy). Subgraph will probably be my live usb OS of choice when it comes out of alpha. Honored to have the president working on part of this issue! :)

**Comment by dma** [comment]
I'll return to it in the next week or so. I'll update this thread when I have a new version. I don't *think* I'm far from a fully reproducible build, which will be pretty awesome.

FWIW, Monero is the only Cryptocurrency project (besides BTC) that interests me at the moment because of its utility and active development/innovation.

**Comment by CameronRuggles** [comment]
>Also, not many people seemed to care

Probably most people don't realize the impact it will have. Getting Monero
in TAILS would be a huge deal. Especially now with multisig and
subaddresses in the next release. It would have an obvious and big impact
on the DNMs, and over all adoption/usage of Monero. Plus, I'd love to just
have it easy to install on my Debian machines.

**Comment by dma** [comment]
Yeah, obviously *I* realize that, it is implied by the earlier remark about the current tangible utility of XMR vs other "privacy" coins, most of which are quite problematic to varying degrees. 

Additionally, with the lack of a hardware wallet so far I would think exploit mitigation / system hardening in @subgraph OS would be something users want on hosts where they use a software wallet? But nobody replied here. We got no feedback at all - zero.

Anyways, as I said, I'll continue working on it and we'll make put it into our repo when our .deb can be reproducibly built.

**Comment by sanecito** [comment]
Yep with the lack of a hardware wallet it would be awesome to use Monero on a more security focused OS! 

Apologies for the lack of testing feedback. 

- What in particular are you looking for feedback, anything beyond reporting issues per the existing doc? 
 - A simple 'instructions worked for me and were clear' from those who install it and system specs (e.g. CPU) so you know some amount of people tested it and what it works on? 
- Would feedback of the 'old' builds still be useful at this point to you or should I hold on asking other community members to test until the new builds? 
- What would be the best way to pass along feedback? Would you prefer comments in this thread or through Subgraph's 'Contact Us'? Open and close an issue for feedback (after official release) on https://github.com/subgraph/subgraph-debian-packages/issues?

I can pass along your requests in the open of the [community meeting in two days](https://github.com/monero-project/meta/issues/155) depending on what you would like. In that way it'd reach the more active members who engage in the meeting or read the logs that get posted, but the community at large likely wouldn't be aware of it and use it until you post your release announcement on r/Monero, etc.

**Comment by dma** [comment]
I dunno, nothing elaborate, maybe just: "I tried it, didn't work", or "I tried it, works". 

Note that one doesn't need SGOS, any Debian derivative would do I think, including Tails.

Not grumbling or anything, I just juggle lots of priority things. This seemed important to some people, and was a longstanding dormant feature request, I picked it up & reported progress, but nobody seemed to care so I put it on backburner (also: holidays). I'll get back on it shortly!

**Comment by dEBRUYNE-1** [comment]
>I dunno, nothing elaborate, maybe just: "I tried it, didn't work", or "I tried it, works".

Note that this ticket is kind of "buried" and therefore not easily visible for potential testers. I'd recommend, once you finalize the new package(s), to put out a post on Reddit (r/monero) which calls for testers. I am sure you'll find a few interested people.

**Comment by dma** [comment]
Obviously, that's why I tweeted about it though to our 7000+ project followers and it was even retweeted by fluffypony. I am not sure even one person tried the .debs. Shrug.

**Comment by Gingeropolous** [comment]
where do i find the .debs?

i dont tweet

**Comment by dma** [comment]
Ah, I see. I should have clarified, Debian stretch derived distro. Thanks for the feedback. Try forcing an install maybe? If it works, we can relax the dependency.

Those debs are a bit out of date. I'm building new packages now & will put them online later. We can move this discussion to avoid adding a lot of noise in here.

**Comment by Gingeropolous** [comment]
indeed, I tend to get vociferous. I'm usually hanging out on freenode #monero

**Comment by dEBRUYNE-1** [comment]
@dma Can you confirm that you updated the binaries here?

https://subgraph.com/sgos/documentation/monero/

**Comment by dma** [comment]
I have now. I just uploaded some packages I built today.

**Comment by Coded-Dude** [comment]
I tried building it for deb with the ubuntu dependencies, but it failed.  I opened an issue before seeing this.  you can close it if you like.....

**Comment by dma** [comment]
Progress update: I've got a monero deb building reproducibly. Monero-gui next, and then we're done.

Build 1 (AWS/Debian 9):

admin@ip-10-0-0-240:~/dev/subgraph-debian-packages$ sha256sum /tmp/build-area/monero_0.11.0.0-2_amd64.deb 
62ad9fb50bb7dc92f6298fccba6a4601a4c7d40fa175396115d1abd16ab90436  /tmp/build-area/monero_0.11.0.0-2_amd64.deb

Build 2 (dedicated box/Debian 9):

dma@spirit:~/dev/subgraph-debian-packages$ sha256sum /tmp/build-area/monero_0.11.0.0-2_amd64.deb 
62ad9fb50bb7dc92f6298fccba6a4601a4c7d40fa175396115d1abd16ab90436  /tmp/build-area/monero_0.11.0.0-2_amd64.deb

Build 3 (local workstation/Subgraph OS):

user@subgraph:~/dev/subgraph-debian-packages$ sha256sum /tmp/build-area/monero_0.11.0.0-2_amd64.deb 
62ad9fb50bb7dc92f6298fccba6a4601a4c7d40fa175396115d1abd16ab90436  /tmp/build-area/monero_0.11.0.0-2_amd64.deb

There some decisions to make for us, like: where does the blockchain data go? Do we start the daemon with systemd by default? Appreciate thoughts on this.

**Comment by sanecito** [comment]
Tried aptitude, but latest package in Synaptic package manager for libssl is 1.0.2g-1ubuntu4.10

> libboost-all-dev is already installed at the requested version (1.58.0.1ubuntu1)
> libboost-all-dev is already installed at the requested version (1.58.0.1ubuntu1)
> No packages will be installed, upgraded, or removed.
> 0 packages upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
> Need to get 0 B of archives. After unpacking 0 B will be used.
>                                          
> libssl-dev is already installed at the requested version (1.0.2g-1ubuntu4.10)
> libssl-dev is already installed at the requested version (1.0.2g-1ubuntu4.10)
> No packages will be installed, upgraded, or removed.
> 0 packages upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
> Need to get 0 B of archives. After unpacking 0 B will be used.

Currently exploring using the libboost 1.62 package from Debian and getting past the Error:Dependency is not satisfiable: libboost-fiber-dev issue

**Comment by diederikdehaas** [comment]
On maandag 22 januari 2018 05:33:37 CET Scott Anecito wrote:
> Tried aptitude, but latest package in Synaptic package manager for libssl is
> 1.0.2g-1ubuntu4.10

On my Debian Stable system, I have both libssl-dev (for ssl 1.1) and 
libssl1.0-dev (for ssl 1.0). They're not co-installable though.
I don't know what/how Mint/Ubuntu has done, but there may be a similar 
situation.

**Comment by hyc** [comment]
> There some decisions to make for us, like: where does the blockchain data go? Do we start the daemon with systemd by default (feeling like no, as it can be started in GUI)? Appreciate thoughts on this.

If you start it from systemd, it should probably have its own uid and a home directory in /var/spool/xxx or something, for storing the blockchain. But it's a good point that it can be started in the GUI. Might be best to leave it at that, thus running under the user's uid and in the user's home directory.

**Comment by danrmiller** [comment]
@dma Can you point me to your debian control files like the buildinfo file?

**Comment by radfish** [comment]
Thanks for packaging!

You can take a look at how it's done in the Arch package:
https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=monero-git
https://aur.archlinux.org/cgit/aur.git/tree/monero.install?h=monero-git

Blockchain data dir is /var/lib/monero which is also the home dir of a dedicated user with that name. It is best for security to have a dedicated user for security isolation purpose, since monerod is internet-facing. Config file is in /etc/monero.conf. Log could go into syslog or /var/log/monero.

Systemd service  exists in the repo. It should be installed, but not sarted automatically, although I think some packages on Debian do start services automatically (not a fan at all, since user should have a chance to edit the config before starting the service).

This setup works fine with the GUI wallet running under the normal user. See monero-wallet-qt package: https://aur.archlinux.org/packages/monero-wallet-qt

I would vote against starting the daemon from the GUI wallet. Having it as a standard systemd service is simple and least surprize to users.

**Comment by dma** [comment]
@radfish @hyc I generally agree with all you're saying, and that's how I did it.

I have a reproducibly built .deb now with .desktop and icons. 

Going to make an apparmor policy for monerod and then call it done as we aren't sandboxing the daemon in Subgraph OS. I believe the packages, which build in Debian 9, should work in Tails. Will update this ticket later on this (cc @danrmiller).

**Comment by CameronRuggles** [comment]
Any updates? Anything I can do to help as a fairly non-technical person?

Super excited for this.

**Comment by agx** [comment]
I have filed a RFP (Request for packaging) in the Debian bug tracker (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=890974).

**Comment by dma** [comment]
Hey all, sorry for lack of updates, it's basically done for Subgraph OS. I'm happy with the package. It builds reproducibly, and I added icons, the desktop file, apparmor profile.. I'll update my last changes to our repo and then we'll write a blog post about it today or tomorrow. It can be built in Debian, though I imagine a maintainer might want to do some things differently.

**Comment by h01ger** [comment]
@dma: where are your source packages? https://subgraph.com/sgos/documentation/monero/ only links to binary packages and both https://github.com/subgraph/monero and https://github.com/subgraph/monero-gui lack a debian/ subdirectory (needed to build source packages)

**Comment by dma** [comment]
@h01ger Check the debian branch of both repositories. That documentation page is out of date, need to update it. I have some commits to the debian branches I haven't pushed yet. Just haven't had time to close this out.

**Comment by h01ger** [comment]
@dma: thanks for the quick reply! and TBH, I was stupid/tired, I looked for branches with "git branch" instead of "git branch -av"... looking forward to your unpushed changes too!

**Comment by dma** [comment]
@h01ger I'm happy with it. I hope Debian/Tails can benefit from it without much extra work. Our build environment pulls packages from the subgraph repo, but there are no subgraph dependencies for building monero and monero-gui (obviously), so those can be removed for a pristine Debian pbuilder build that's reproducible. I'll push the apparmor policy I made for monerod and whatever else is still not online & then I think we're (I'm) done..

**Comment by h01ger** [comment]
@dma: the debian packaging looks very nice, kudos! (and many thanks too)

I just saw one instance of subgraphos specific changes, the systemd service requires the subgraph_metaproxy.service - so now I wonder how you would deal if I'd uploaded that package to Debian proper, without that requirement...

(and do take your time, I wont really do something within the next 24h anyway. but then, very probably!)

**Comment by h01ger** [comment]
only monero/external/ seems a bit scary...

**Comment by dma** [comment]
@h01ger yeah, right, in the systemd file is a subgraph dependency, which you won't want (I may even remove that - I just wanted to make sure monerod doesn't start connecting to the Internet until transparent tor-ification is up), forgot about that. If Debian packages Monero/Monero GUI, I imagine we'll drop ours and use Debian's. If we need to overload the systemd unit file, we can do that separately.

**Comment by sanecito** [comment]
Hey folks, wishing to follow up on the great work you're doing packaging Monero for Debian. Has Monero v0.12 been packaged for Debian by chance given the documentation references v0.11 still? It's not much unfortunately, but I did send some Moneroj to SubGraph as thanks for prior work.

**Comment by jonassmedegaard** [comment]
Monero entered Debian unstable earlier today. Health of the packaging can be monitored at <https://tracker.debian.org/pkg/monero>

**Comment by spth** [comment]
Any news on progress getting a GUI into Debian?

Philipp

**Comment by sanecito** [comment]
@spth The best point of contact is probably going to be @rehrar as he is the point person between Monero and Purism's for the librem5 integration efforts. If you're unfamiliar with librem5 it's basically a phone that will run PureOS (Debian based) by default and is launching April 2019, so there is interest from those outside the Monero community in contributing to this issue.. It looks like currently we now have a test package for Debian unstable/sid: https://packages.debian.org/search?keywords=monero

I'll ask around if anyone knows of any taskboards, etc to better track how close we are to a non-test package.

**Comment by jonassmedegaard** [comment]
Quoting Scott Anecito (2019-01-08 19:14:04)
> @spth The best point of contact is probably going to be @rehrar as he is the point person between Monero and Purism's for the librem5 integration efforts. If you're unfamiliar with librem5 it's basically a phone that will run PureOS (Debian based) by default and is launching April 2019, so there is interest from those outside the Monero community in contributing to this issue.. It looks like currently we now have a test package for Debian unstable/sid: https://packages.debian.org/search?keywords=monero
> 
> I'll ask around if anyone knows of any taskboards, etc to better track how close we are to a non-test package.

I guess you want this: https://tracker.debian.org/pkg/monero

Also, I maintain the monero package officially in debian, and I work for 
Purism with one of my tasks being point person towards Monero.

Feel free to reach out to me at either jonas.smedegaard@puri.sm or 
dr@jones.dk as needed :-)

 - Jonas

-- 
 * Jonas Smedegaard - idealist & Internet-arkitekt
 * Tlf.: +45 40843136  Website: http://dr.jones.dk/

 [x] quote me freely  [ ] ask before reusing  [ ] keep private

**Comment by jonassmedegaard** [comment]
> Any news on progress getting a GUI into Debian?

No progress on packaging the GUI part of Monero yet, unfortunately.

**Comment by who-biz** [comment]
I began packaging the GUI for Debian.  Will definitely take some more work to get the fine-tuning of everything in order... but the repository at https://github.com/who-biz/monero-gui/tree/debian is a start.  It does build from source cleanly with `dpkg-buildpackage -G` issued from the source directory. (You will need to run that command as root to package -- do so from within a chroot, or a secure environment)

All of the relevant files are contained in the `debian` directory. While it is still a WIP, it builds a functioning .deb package for release v0.13.0.4.

Anyone can feel free to contribute via PR or just appropriate what is there for your own use, as you see fit.  Just trying to get the ball rolling on this, so copy-and-pasting to a different repo, or whatever anybody may want to do, is completely okay by me.

**Comment by adrelanos** [comment]
"simply put the pre-build Monero binaries into a deb package": [[IDEA] [PROPOSAL] Monero Debian (deb) packages / Debian package repository deb.getmonero.org (I can do)](https://www.reddit.com/r/Monero/comments/cowjun/idea_proposal_monero_debian_deb_packages_debian/)

**Comment by denisgoddard** [comment]
The [excuses](https://qa.debian.org/excuses.php?package=monero) for this .deb say it's stuck because we need a clean arm64 build. Bug [939311](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=939311) is open for this problem.

Is anybody actively working that Bug? If not, I'll get an arm64 compute from a random cloud service and try building, I guess...

**Comment by bill-mcgonigle** [comment]
> Is anybody actively working that Bug? If not, I'll get an arm64 compute from a random cloud service and try building, I guess...

@denisgoddard - it looks like [the blocker bug](https://sourceware.org/bugzilla/show_bug.cgi?id=25210) is fixed as of last month; are you on this?

**Comment by sanecito** [comment]
There is now a CCS that would address the heart of this issue of making it easier for users to install the Monero GUI on Debian related distros courtesy of Whonix folk and infrastructure. If people are still interested in taking the path of getting the package onto getmonero.org or Debian infrastructure that would be cool too.

If so, let me know so I can update this issues scope to package being hosted on getmonero and/or Debian so work can still be tracked, etc.

https://repo.getmonero.org/monero-project/ccs-proposals/-/merge_requests/130?commit_id=b188c28babbaa21f12a3f4993d8ed046b675fcaf

**Comment by bill-mcgonigle** [comment]
OK, sweet, Patrick has done the packaging:
https://gitlab.com/whonix/monero-gui

Not sure if it meets all of Debian's requirements but it's at least that much closer.

**Comment by jonassmedegaard** [comment]
Quoting Bill McGonigle (2020-08-14 16:40:06)
> OK, sweet, Patrick has done the packaging:
> https://gitlab.com/whonix/monero-gui
> 
> Not sure if it meets all of Debian's requirements but it's at least that much closer.

It certainly does not meet Debian's requirements to put the pre-build 
Monero binaries into a deb package - but if acceptable to some then good 
for them!


 - Jonas

-- 
 * Jonas Smedegaard - idealist & Internet-arkitekt
 * Tlf.: +45 40843136  Website: http://dr.jones.dk/

 [x] quote me freely  [ ] ask before reusing  [ ] keep private

**Comment by erciccione** [comment]
Posting this here since it's related:

I recently opened an issue on Tails's repository asking to add Monero. That's not possible at the moment, but they give some good suggestions about getting the GUI on Debian: https://gitlab.tails.boum.org/tails/tails/-/issues/17823

**Comment by adrelanos** [comment]
> OK, sweet, Patrick has done the packaging:
> https://gitlab.com/whonix/monero-gui
> 
> Not sure if it meets all of Debian's requirements but it's at least that much closer.

No, this was explicitly excluded. Quote [CCS proposal](https://ccs.getmonero.org/proposals/adrelanos-debian-package.html):

> `packages.debian.org` is out of scope since Debian does not upgrade often enough to keep up with Monero fork cycle and for other reasons as well.

https://repo.getmonero.org/monero-project/ccs-proposals/-/merge_requests/130

**Comment by erciccione** [comment]
Tails' devs posted a list of requirements needed in order to support Monero on tails: https://gitlab.tails.boum.org/tails/tails/-/issues/17823#note_163991

It's fundamental that the package maintained by the 'Debian Cryptocoin Team' (https://tracker.debian.org/pkg/monero) is fixed and updated to the latest version before Debian Bullseye's soft freeze (December 12th. In less than a month). Recently, an unfixed bug caused the monero package to be removed from 'testing' and that's problematic for a future addition on Tails (and not only that, please see the gitlab issue i linked to at the beginning of this post).

So, to reiterate, the 'monero' package needs to be fixed and updated ASAP. If anybody has contacts with the developers who used to work on this, please contact them and point them to this issue.

**Comment by erciccione** [comment]
The issue i mentioned above is resolved. The latest Monero release is now available on Debian testing.

See https://gitlab.tails.boum.org/tails/tails/-/issues/17823#note_164543

**Comment by DeeDeeRanged** [comment]
Any chance to get the monero package updated to the latest version for Debian testing/sid?

**Comment by selsta** [comment]
https://tracker.debian.org/pkg/monero
